compile/ivy/src/main/scala/sbt/compiler/ComponentCompiler.scala

```diff
@@ -23,12 +25,40 @@ object ComponentCompiler {
         componentCompiler(compilerInterfaceID)
       }
   }
+
+  def interfaceProvider(manager: ComponentManager, ivyConfiguration: IvyConfiguration): CompilerInterfaceProvider = new CompilerInterfaceProvider {
+    def apply(scalaInstance: xsbti.compile.ScalaInstance, log: Logger): File =
+      {
+        // this is the instance used to compile the interface component
+        val componentCompiler = new IvyComponentCompiler(new RawCompiler(scalaInstance, ClasspathOptions.auto, log), manager, ivyConfiguration, log)
+        log.debug("Getting " + compilerInterfaceID + " from component compiler for Scala " + scalaInstance.version)
+        componentCompiler(compilerInterfaceID)
+      }
+  }
+
+  /**
+   * Generates a sequence of version of numbers, from more to less specific from `fullVersion`.
+   */
+  private[compiler] def cascadingSourceModuleVersions(fullVersion: String): Seq[String] = {
+
+    Try {
+      val VersionNumber(numbers @ maj +: min +: _, tags, _) = VersionNumber(fullVersion)
+      Seq(fullVersion,
+        numbers.mkString(".") + (if (tags.nonEmpty) tags.mkString("-", "-", "") else ""),
+        numbers.mkString("."),
+        s"$maj.$min"
+      ).distinct
```

Comment:
I implemented it directly in `VersionNumber`, and moved the tests to `VersionNumberSpec`. The implementation is much shorter now :).